### PR TITLE
Polling cluster formation state for master-is-stable health indicator

### PR DESCRIPTION
**Pull request by masseyke**
As shown in step 1.2.2.4 in the diagram at https://github.com/elastic/elasticsearch/issues/85624#issuecomment-1175421820, if there has not been a node elected master for more than 30 seconds, then when a user queries a master node it is supposed to reach out to all other master-eligible nodes and summarize the results of their cluster formation knowledge. Since we don't want to block a user request while making those remote calls, it should be done speculatively so that we have the information when it is needed. If a master-eligible node gets a cluster state changed event that the master is null, then it will wait 10 seconds (in case it gets another cluster state changed event that the master has flipped to non-null quickly, which is the most likely scenario), then reach out to all of the other master-eligible nodes using a transport action (#87306) to get their cluster formation data. This data is stashed away in case a user request comes in. Since the user won't hit this path for 30 seconds after a master has become null (because we tell the user they have a stable master unless it's been missing for more than 30 seconds), we have plenty of time for this -- 10 seconds of delay in case we're notified that there is a new master plus 10 seconds to get the result back from the master-eligible node before timing out, plus 10 seconds to spare.
Since it includes nested callbacks, this code is a little hard to follow. Here's the outline:
For each master-eligible node in the cluster, we 

1. Schedule a task to be performed after 10 seconds that connects to the master-eligible node and returns a handle to a Cancellable. This returns immediately because all of the work is done asynchronously.
    1. On an asynchronous successful response from that call to connect, we send the transport request. This also returns immediately because its work is done asynchronously.
        1. On an asynchronous successful response from the transport request, we close the connection update the shared map with the result. Then we start back at step 1 again.
        2. On an asynchronous failure from the transport request,  we close the connection and update the shared map with an exception. Then we start back at step 1 again.
    2. On an asynchronous failure from connect, we update the shared map with an exception and then start back at step 1 again.

This PR was originally a part of #88020, but has been broken off to simplify the discussion. This PR only populates the `CoordinationDiagnosticsService.clusterFormationResponses` field, which will be used by an updated #88020 to drive the master stability logic.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java [hunk 1]
```diff
@@ -410,6 +441,200 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+
+    /**
+     * This class represents a collection of related Cancellables. If one is cancelled, they are all considered cancelled. If cancel() is
+     * called on this method, then cancel() is called on all child Cancellables.
+     */
+    private static class MultipleCancellablesWrapper implements Scheduler.Cancellable {
+        private final List<Scheduler.Cancellable> delegates;
+
+        MultipleCancellablesWrapper() {
+            this.delegates = Collections.synchronizedList(new ArrayList<>());
```
Comment (masseyke) [inline server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:488]: Ignore this for now. As it is, I think it could blow up when we're iterating through this list. I'm not fixing it for the moment since I've sent out an email with line numbers.

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java [hunk 2]
```diff
@@ -410,6 +441,204 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+    // Non-private for testing
+    class PollClusterFormationStateTask {
+        /**
+         * The node that is being polled
+         */
+        private final DiscoveryNode node;
+        /**
+         * This is a reference to the global nodeToClusterFormationStateMap that was current at the time this object was constructed. The
+         * global map is recreated whenever the task is cancelled. Having this reference prevents accidental writes to that map after
+         * cancellation.
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:493]: This is a bit confusing as it implies we're copying it (ie "that was current at the time this object was constructed"), but also "references" it ... which means it's going to change in line with the global map. 

I'm also not sure about the accidental writes to "that" map (it's "this" map isn't it :) ?) nor what's accidental about the said writes.

Could we maybe talk about the why we have this map here as opposed to the implementation details of how it arrived here? (apologies if I'm misunderstanding its purpose) 

Update: maybe we should drop `PollClusterFormationStateTask` altogether, in which case ignore the above :)

**Review by andreidan — COMMENTED** [review]
Thanks for working on this Keith.

I've left a few suggestions to (hopefully) simplify it a bit

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java [hunk 3]
```diff
@@ -410,6 +441,204 @@ public void clusterChanged(ClusterChangedEvent event) {
...
         }
+        if (currentMaster == null && clusterService.localNode().isMasterNode()) {
+            /*
+             * This begins polling all master-eligible nodes for cluster formation information. However there's a 10-second delay before it
+             * starts, so in the normal situation where during a master transition it flips from master1 -> null -> master2, it the
+             * polling tasks will be canceled before any requests are actually made.
+             */
+            beginPollingClusterFormationInfo();
+        } else {
+            cancelPollingClusterFormationInfo();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:452]: It's not a huge deal. Maybe we could use `null` instead of `new ConcurrentHashMap<>()` to distinguish "we're not polling" from "we're polling but have no entries"?

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java [hunk 4]
```diff
@@ -410,6 +441,204 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+            this.multipleCancellablesWrapper = multipleCancellablesWrapper;
+        }
+
+        /**
+         * This method returns a Cancellable quickly, but in the background schedules to query the remote node's cluster formation state
+         * in 10 seconds, and repeats doing that until cancel() is called on the returned Cancellable.
+         *
+         * @return
+         */
+        public Scheduler.Cancellable pollUntilCancelled() {
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:533]: Oh good point about PollClusterFormationStateTask's existence. In an earlier incarnation this was an actual Runnable, and `pollUntilCancelled()` was just `run()`. I improved that, but forgot to move the state out of `PollClusterFormationStateTask` and then get rid of it altogether. I think the signature will have to be a little different though. More on that in an upcoming comment.

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java (hunk 2 above)
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:493]: Below is a scenario showing why I'm copying the `clusterFormationResponses` reference to and passing it in rather than using the global `clusterFormationResponses`. Maybe I'm being unnecessarily cautious.

We begin polling master eligible nodes A, B, C
  We create a new global clusterFormationResponses map
  We update the global clusterFormationResponses with results from A, B
We cancel polling
We begin polling master eligible nodes  A, B, D (note that it's a different set)
  We create a new global clusterFormationResponses map
  We update the global clusterFormationResponses with results from C (the write was happening before cancel took effect and I don't think that it will interrupt a map insert but maybe I'm wrong)
  We update the global clusterFormationResponses with results from A, B, D

So now we have result for A, B, C, and D, even though C is no longer part of the cluster. So we get weird results when we look to see if we can form a quorum, or if C knows about the existence of D. Or even if it's not as extreme as a node being replaced, we have results that might be out of date, causing confusion. My simple way to avoid worrying about all of this was to just pass the method a reference to the clusterFormationResponses map that it is supposed to be using.

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java]: Here's the signature I have now:
`void fetchClusterFormationInfo(DiscoveryNode node, ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> nodeToClusterFormationStateMap, List<Scheduler.Cancellable> cancellables)`
I can't just return a single Cancellable because one call to the method can generate lots of Cancellables (since the method schedules calls of itself). So I'm passing in the array of Cancellables that the `cancelPollingClusterFormationInfo` acts on. This avoids having to have the `MultipleCancellablesWrapper` but it effectively does the same thing. And I passed in a nodeToClusterFormationStateMap because I can't have just a ClusterFormationInfoAction.Response Consumer because one possible outcome is an Exception rather than a Response. I'll see if making that a Consumer<ClusterFormationStateOrException> makes it more or less readable.

[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java]: OK the method now accepts a `Consumer<ClusterFormationStateOrException>` instead of a `ClusterFormationStateOrException> nodeToClusterFormationStateMap`. I think it helps out a little bit. The functionality is unchanged.

[the remark this answers is not in the document]
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java]: I think this is a step in the right direction, thanks for implementing it ! 

>I can't just return a single Cancellable because one call to the method can generate lots of Cancellables (since the method schedules calls of itself)

Would this be an avenue for simplification?

Say the method signature is as follows
```
Scheduler.Cancellable fetchClusterFormationInfo(DiscoveryNode node, Consumer<ClusterFormationInfoAction.Response> responseConsumer)
```
This means the caller is responsible for re-calling `fetchClusterFormationInfo` method when the `responseConsumer` is called. 

I think this would remove the need for the `runAfter` call in that `runAfter(runBefore` sequence because it seems we'll want to reschedule irrespective of being able to get the response or not - so when `responseConsumer` is called we should be rescheduling. 
ie. we always call `responseConsumer` irrespective of how `fetchClusterInfoListener` completes:
```
                new ActionListenerResponseHandler<>(
                    ActionListener.runBefore(fetchClusterInfoListener, () -> Releasables.close(releasable)),
                    ClusterFormationInfoAction.Response::new
                )
```

Consuming the `fetchClusterFormationInfo` would look something along the lines of:

```   
        getMasterEligibleNodes().forEach(masterNode -> {
            Consumer<ClusterFormationStateOrException> responseConsumer = response -> {
                if (clusterFormationResponses != null) {
                    clusterFormationResponses.put(masterNode, response);
                }
            };
            
            Scheduler.ScheduledCancellable scheduleFetch = fetchClusterFormationInfo(masterNode, responseConsumer.andThen(response -> {
                if (clusterFormationInfoTasks != null) {
                    // reschedule the fetch if it wasn't cancelled already
                    clusterFormationInfoTasks.add(fetchClusterFormationInfo(masterNode, responseConsumer));
                }
            }));
            
            clusterFormationInfoTasks.add(scheduleFetch);
        });

```

This would reduce the scope of `fetchClusterFormationInfo` and the need for a `Consumer<Cancellable>`.

What do you think?

[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java]: Unless I'm misunderstanding something, we can't use a `Consumer<ClusterFormationInfoAction.Response>` here because then we'd have no way of handling exceptions, right? And we need to keep track of exceptions we encounter.
And even using a `Consumer<ClusterFormationStateOrException>` I don't think we can take the approach you're suggesting will work as-is because it misses retries on a `connectToNode` exception (https://github.com/elastic/elasticsearch/pull/88397/files#diff-6898ba5666bbb351e267e9295e8feb7fbabf0d95d77aa4375c06cb2c16ad0a7dR529). I could probably add in a separate consumer for `connectToNode`, but at that point it's starting to get even more complex. Is there any use case that is missed by the current code?

[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java]: Also I think your proposal exposes us to the race condition here right (although I'm not 100% sure that causes big problems)? https://github.com/elastic/elasticsearch/pull/88397#discussion_r919348039
But that is fixable.

[the remark this answers is not in the document]
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java]: >And even using a Consumer<ClusterFormationStateOrException> I don't think we can take the approach you're suggesting will work as-is because it misses retries on a connectToNode exception (https://github.com/elastic/elasticsearch/pull/88397/files#diff-6898ba5666bbb351e267e9295e8feb7fbabf0d95d77aa4375c06cb2c16ad0a7dR529)

Ah apologies I did mean `Consumer<ClusterFormationStateOrException>`. But if the `connectToNode` raises an exception the response consumer will still be called https://github.com/elastic/elasticsearch/pull/88397/files#diff-6898ba5666bbb351e267e9295e8feb7fbabf0d95d77aa4375c06cb2c16ad0a7dR524 (and the responseConsumer will track the exception and reschedule the `fetch`). Am I misisng something? 

> Is there any use case that is missed by the current code?

I think it's a bit difficult to follow, that's why I suggested reducing the scope of `fetch...` and have a clear view over which callback reschedules the polling and when that happens

**Comment by masseyke** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java [hunk 5]
```diff
@@ -410,6 +444,168 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+            );
+        });
+    }
+
+    /**
+     * This wraps the responseConsumer in a Consumer that will run beginPollingClusterFormationInfo() after responseConsumer has
+     * completed, adding the resulting Cancellable to cancellableConsumer.
+     * @param masterEligibleNode The node being polled
+     * @param responseConsumer The response consumer to be wrapped
+     * @param cancellableConsumer The list of Cancellables
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:502]: this is not accurate anymore

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java [hunk 6]
```diff
@@ -410,6 +444,168 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+                fetchClusterFormationInfo(
+                    masterEligibleNode,
+                    responseConsumer.andThen(rescheduleFetchConsumer(masterEligibleNode, responseConsumer, cancellableConsumer))
+                )
+            );
+        });
+    }
+
+    /**
+     * This wraps the responseConsumer in a Consumer that will run beginPollingClusterFormationInfo() after responseConsumer has
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:498]: it will run `fetchClusterFormationInfo` I believe?

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsServiceTests.java [hunk 7]
```diff
@@ -586,6 +593,115 @@ public void testRedForNoMasterAndWithMasterEligibleNodesAndNoLeader() throws IOE
...
+                nodeToDisconnect.heal();
+            });
+        }
+    }
+
+    public void testBeginPollingClusterFormationInfoCancel() {
+        /*
+         * This test sets up a 4-node cluster (3 master eligible). We call beginPollingClusterFormationInfo() on each node. We then
+         * cancel all tasks. This simulates what will happen most often in practice -- polling is triggered when the master node goes
+         * null, and then polling is cancelled when a new master node is elected within 10 seconds. We then simulate the cluster running
```
Comment (andreidan) [inline server/src/test/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsServiceTests.java:678]: "polling is cancelled when a new master node is elected within 10 seconds"

I think polling is cancelled immediately after a master node is detected. Maybe I'm misunderstanding what you meant here by the 10 seconds delay?

**Review by andreidan — APPROVED** [review]
Thanks for iterating on this Keith

LGTM
